### PR TITLE
feat: configurable OAuth scopes via config.toml (#184)

### LINKED ISSUE #184: feat: make OAuth scopes configurable via config.toml
## Problem

The OAuth 2.0 flow hardcodes classic scopes in `src/api/auth.rs`:

```rust
const SCOPES: &str = "read:jira-work write:jira-work read:jira-user offline_access";
```

Users who create an OAuth app in the Atlassian Developer Console with granular scopes (e.g., `write:comment:jira`, `write:issue:jira`) cannot use them — `jr auth login --oauth` always requests the hardcoded classic scopes regardless of what the app is configured for.

This matters for LLM/agent use cases where least-privilege is critical. `write:jira-work` grants destructive permissions (delete worklogs, remove links, transition to any state) that an agent should not have. Atlassian offers granular scopes specifically for this purpose, but jr can't request them.

## Expected

Read OAuth scopes from `config.toml` when present, fall back to the current hardcoded default for backwards compatibility:

```toml
[instance]
auth_method = "oauth"
oauth_scopes = "read:jira-work read:jira-user write:issue:jira write:comment:jira write:issue-link:jira offline_access"
```

In `src/api/auth.rs`, `oauth_login()` would accept scopes as a parameter instead of using the constant:

```rust
pub async fn oauth_login(client_id: &str, client_secret: &str, scopes: &str) -> Result<OAuthResult> {
```

If `oauth_scopes` is not set in config, use the current `SCOPES` constant — no behavior change for existing users.

## Impact

Users building agent tooling (MCP servers, CLI wrappers, AI coding assistants) currently have two options: accept broad classic scopes, or fork the repo to change one line. Neither is sustainable. A config-driven approach lets each user tailor scopes to their OAuth app without code changes.

## Related

- ADR-0002: OAuth approach was superseded to require user-provided OAuth app credentials — this is the natural next step (user-provided scopes alongside user-provided credentials)
- #183: verbose mode gaps make debugging scope mismatches harder

### PR DESCRIPTION
**Pull request by Zious11**
## Summary

Makes OAuth 2.0 scopes configurable via `[instance].oauth_scopes` in `config.toml`, falling back to the existing hardcoded classic scope set when unset. Unlocks granular / least-privilege OAuth apps for LLM and agent use cases (Atlassian Developer Console apps configured with scopes like `write:issue:jira` instead of the broader `write:jira-work`).

Zero behavior change for users who don't set the new key.

## Scope-model constraints (validated with Perplexity against Atlassian's developer docs)

- Classic scopes remain Atlassian-recommended for Jira Platform apps; granular is for cases "when you can't use classic scopes." Jira Software is the exception (granular-only).
- Classic and granular scopes **cannot mix** in one `/authorize` request — the scope list must match what the Developer Console app has configured.
- `offline_access` is required in both modes for a refresh token to be issued.

## Key design decisions

- Pure passthrough — no client-side scope-mix detection. Atlassian returns `invalid_scope` at token exchange if the user's config is wrong; maintaining a client-side allowlist of valid scope combinations would be brittle as Atlassian revises the list.
- `refresh_oauth_token` is unchanged — the `refresh_token` grant inherits scopes from the original authorization per RFC 6749 §6. Users must re-run `jr auth login --oauth` for changed scopes to take effect (documented on the function).
- Env-var override (`JR_INSTANCE_OAUTH_SCOPES`) is out of scope for this PR — figment's default `Env::prefixed("JR_")` can't reach nested struct fields, and the separator fix (`.split("_")` or `.split("__")`) is orthogonal scope. Follow-up issue to track proper nested-env support.

## Changes

- `src/config.rs`: add `InstanceConfig.oauth_scopes: Option<String>` + 2 TOML-parsing unit tests
- `src/api/auth.rs`: rename `const SCOPES` → `pub const DEFAULT_OAUTH_SCOPES`; add `scopes: &str` parameter to `oauth_login`; document scope-inheritance on `refresh_oauth_token`
- `src/cli/auth.rs`: module-private `resolve_oauth_scopes(&Config) -> Result<String>` (None → default, valid → trimmed + whitespace-collapsed, empty / whitespace-only → `ConfigError` exit 78); `login_oauth` threads it into `oauth_login`; 6 unit tests including a content pin on `DEFAULT_OAUTH_SCOPES` and a passthrough lock-in covering classic+granular mix / missing `offline_access` / unknown scopes
- `README.md`: document the new key with the three constraints users need to know
- `docs/specs/oauth-scopes-configurable.md` + `docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md`: spec + TDD plan

## Deferred to follow-up issues

- #255 — OAuth `state` parameter uses wall-clock nanoseconds instead of CSPRNG (pre-existing, not introduced by this PR)
- #256 — `client_id` / `redirect_uri` / `state` not URL-encoded in authorize URL; only `scope` is (defense-in-depth)

## Test plan

- [x] `cargo fmt --all -- --check` clean
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo test` — full suite passes; 8 new unit tests (2 in `config.rs`, 6 in `cli/auth.rs`) all green
- [x] Local multi-agent review (code / security / architecture / test) — 2 rounds, all clean on round 2
- [ ] Copilot review iterate-until-clean

Closes #184.

### CODE REVIEW

File: src/cli/auth.rs [hunk 1]
```diff
@@ -179,9 +202,10 @@ pub async fn login_oauth(
     // Store OAuth app credentials in keychain
     crate::api::auth::store_oauth_app_credentials(&client_id, &client_secret)?;
 
-    let result = crate::api::auth::oauth_login(&client_id, &client_secret).await?;
-
     let mut config = Config::load().unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/cli/auth.rs:206]: Not fixing in this PR — this is pre-existing behavior. The `Config::load().unwrap_or_default()` pattern was already at `src/cli/auth.rs:184` before this PR; the change only moved it a few lines to support threading scopes through `oauth_login`.

Agreed the concern is real (reproducibly overwrites malformed config with defaults after a successful OAuth login), so I filed #258 to track a focused fix that also audits the pattern across the rest of `src/cli/`. Keeping this PR scoped to the OAuth-scopes feature.

File: src/api/auth.rs [hunk 2]
```diff
@@ -142,12 +148,21 @@ pub struct OAuthResult {
...
+///
+/// `scopes` is a space-separated scope string (URL-encoded internally).
+/// Callers should use [`DEFAULT_OAUTH_SCOPES`] when no user override is set.
+/// Note: [`refresh_oauth_token`] does NOT take a scope parameter — the
+/// `refresh_token` grant inherits scopes from the original authorization.
+pub async fn oauth_login(
+    client_id: &str,
+    client_secret: &str,
+    scopes: &str,
+) -> Result<OAuthResult> {
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/api/auth.rs:165]: Current behavior is correct — `jr` is a CLI binary, not a library crate published for external consumption. `src/lib.rs` exists only so integration tests in `tests/*.rs` can reuse internal types (per `CLAUDE.md`: "lib.rs: Crate root (re-exports for integration tests)"), and `Cargo.toml` has `[[bin]]` with no registry-publish metadata. There are no external crate consumers to break.

If we ever did decide to publish `jr` as a library on crates.io, the right move at that point would be a dedicated `-sys`/lib crate split, not compatibility shims on internal helpers. Accepting the signature change here.

File: docs/specs/oauth-scopes-configurable.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# OAuth Scopes Configurable via `config.toml`
+
+**Issue:** [#184](https://github.com/Zious11/jira-cli/issues/184)
+
+## Problem
+
+`jr`'s OAuth 2.0 (3LO) flow at `src/api/auth.rs:27` hardcodes a classic scope string:
+
+```rust
+const SCOPES: &str = "read:jira-work write:jira-work read:jira-user offline_access";
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/specs/oauth-scopes-configurable.md]: Fixed in ed61e4f. Spec now phrases the pre-existing constant historically ('previously hardcoded a classic scope string') and explicitly notes that the PR renames `SCOPES` → `pub const DEFAULT_OAUTH_SCOPES`. Kept `const SCOPES` in the historical snippet since that IS what was in the code before the PR.

File: docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+## File map
+
+| Path | Change |
+|---|---|
+| `src/config.rs` | Add `oauth_scopes: Option<String>` to `InstanceConfig`; add unit tests for TOML parse, None default, env override |
+| `src/api/auth.rs` | Rename `const SCOPES` → `pub const DEFAULT_OAUTH_SCOPES`; add `scopes: &str` parameter to `oauth_login`; use the parameter in the authorize URL |
+| `src/cli/auth.rs` | Add module-private `resolve_oauth_scopes(&Config) -> Result<String>`; call it in `login_oauth` and pass the result into `oauth_login`; add unit tests covering four cases |
+| `README.md` | Add `oauth_scopes` to the global config example and a short note on classic-vs-granular + `offline_access` |
+
+No files created. No integration test file needed — `oauth_login` hits `auth.atlassian.com` directly and is out of scope for unit-testable mocking.
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md]: Fixed in ed61e4f. Two changes to the plan's file map section:

- Removed the stale "env override" test from the `src/config.rs` row (env-var override was moved out of scope during local review; the spec now documents the figment-nesting limitation at `docs/specs/oauth-scopes-configurable.md:50`).
- Rewrote the "No files created" sentence to distinguish **implementation** files (none created — we only modify existing files) from the **documentation** files of this PR (the spec + this plan itself), which are new.

File: src/cli/auth.rs [hunk 5]
```diff
@@ -179,9 +202,10 @@ pub async fn login_oauth(
...
+    let scopes = resolve_oauth_scopes(&config)?;
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/cli/auth.rs]: Fixed in ea85857. Reordered `login_oauth` so `Config::load()` and `resolve_oauth_scopes(&config)?` run **before** `store_oauth_app_credentials(&client_id, &client_secret)`. An empty/whitespace-only `[instance].oauth_scopes` now fails fast with `JrError::ConfigError` (exit 78) without leaving new credentials in the keychain. Added an inline comment at the new call site explaining why the ordering matters.

File: docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+pub struct InstanceConfig {
+    pub url: Option<String>,
+    pub cloud_id: Option<String>,
+    pub org_id: Option<String>,
+    pub auth_method: Option<String>,
+    pub oauth_scopes: Option<String>,
+}
+```
+
+No other changes required — `Deserialize`, `Default`, and the existing `Env::prefixed("JR_")` merge pick this up automatically (empirically confirmed: `tests/auth_refresh.rs:71` already uses `JR_INSTANCE_AUTH_METHOD` against the same struct).
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md]: Fixed in 9bf4265. Removed the "Env::prefixed("JR_") picks this up automatically" claim and replaced with an explicit "do not assume JR_INSTANCE_OAUTH_SCOPES works" caveat pointing to the spec's data-flow note (`docs/specs/oauth-scopes-configurable.md:50`). The `tests/auth_refresh.rs:71` citation is also dropped — it was defensive env-clearing, not proof of propagation, as you noted.

File: docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+- [ ] `cargo test`
+
+```bash
+git add src/config.rs
+git commit -m "feat(config): add oauth_scopes to InstanceConfig (#184)
+
+Adds Option<String> field so users can override the hardcoded OAuth 2.0
+scope list via config.toml or the JR_INSTANCE_OAUTH_SCOPES env var. No
+behavior change yet — the field is unused until Task 3 wires it into
+oauth_login. Spec: docs/specs/oauth-scopes-configurable.md."
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/superpowers/plans/2026-04-22-oauth-scopes-configurable.md]: Fixed in 9bf4265. Dropped the `JR_INSTANCE_OAUTH_SCOPES` mention from the sample Task 1 commit message and added an explicit "env-var override is out of scope for this PR" note referencing the spec.

File: docs/specs/oauth-scopes-configurable.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+Zero behavior change for any user who doesn't set the new config field.
+
+### Data flow
+
+```
+Config::load() → cli::auth::handle_login()
+  → resolve_oauth_scopes(&config)?
+  → api::auth::oauth_login(client_id, client_secret, scopes)
+    → urlencoding::encode(scopes) in the authorize URL
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/specs/oauth-scopes-configurable.md:72]: Fixed in 9bf4265. Data-flow diagram now references `cli::auth::login_oauth()` (the actual function at `src/cli/auth.rs:150`). `handle_login` never existed; the real dispatch is `src/main.rs` → `AuthCommand::Login { oauth: true }` → `login_oauth()`.

File: docs/specs/oauth-scopes-configurable.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+| `oauth_scopes = ""` or whitespace-only | `JrError::ConfigError`, actionable message, exit 78 |
+| Invalid scope combination (e.g. mix, unknown scope) | Atlassian returns `invalid_scope` at token exchange; existing `anyhow::bail!("Token exchange failed: {body}")` path surfaces it |
+
+No client-side mix detection. No warning for missing `offline_access`. Both are documented caveats, not enforcement.
+
+### Documentation
+
+- `docs/specs/oauth-scopes-configurable.md` (this file)
+- README section on OAuth configuration: call out that the scope string must match what the Developer Console app has configured, that classic and granular cannot mix, and that `offline_access` is required for unattended re-authentication
+- Existing `jr auth login --oauth` help text: reference the config key so users discover it without reading the README
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/specs/oauth-scopes-configurable.md]: Fixed in 9bf4265 by actually updating the CLI help text. Added the scope-override note to the doc comment on `AuthCommand::Login.oauth` in `src/cli/mod.rs` so `jr auth login --help` now surfaces the config key. Kept the spec's "update the help text" promise intact rather than scoping it out.
